src/python/pants/backend/tools/semgrep/rules.py

```diff
@@ -85,24 +68,35 @@ def ancestor_configs(self, address: Address) -> Iterable[PurePath]:
             yield from self.configs_by_dir.get(ancestor, [])
 
 
-def _group_by_semgrep_dir(all_paths: Paths) -> AllSemgrepConfigs:
+def _group_by_semgrep_dir(config_paths: tuple[str, ...], all_paths: Paths) -> AllSemgrepConfigs:
     configs_by_dir = defaultdict(set)
     for path_ in all_paths.files:
         path = PurePath(path_)
-        # A rule like foo/bar/.semgrep/baz.yaml should behave like it's in in foo/bar, not
-        # foo/bar/.semgrep
-        parent = path.parent
-        config_directory = parent.parent if parent.name == _RULES_DIR_NAME else parent
+        # Rules like foo/bar/.semgrep/baz.yaml and foo/bar/.semgrep/baz/qux.yaml should apply to the
+        # project at foo/bar
+        config_directory = path.parent
+        for config_path in config_paths:
+            if config_path not in path.parts:
+                continue
+            config_directory = PurePath(*path.parts[: path.parts.index(config_path)])
+            break
```

Comment:
This nested loop jumps out to me as a bit concerning, due to performance:

- this loop was previously approximately O(`len(all_paths)`)
- I think it's now O(`len(all_paths)` * average number of `parts` of each path * `len(config_paths)`) = O(`sum(len(path.parts) for path in all_paths)` *  `len(config_paths)`)
 
That is, it's now proportional to the number of config files _and_ the number of parts (e.g. `a`, `b` etc. in `a/b/...`) for all files in the whole repository. I imagine the number of config files is often 1, but could easily be more (with a `.semgrep` directory)... but the number of path components could be very high, depending how deeply nested the repo is!

I haven't yet thought of any alternatives, and how much effort to spend here will be guided by the top-level conversation about globbing etc.

Reply:
Good point. 

I'm comfortable forcing just one configuration option here. 

For the iteration, I was really hoping for glob utils that could give you paths as match groups based on the nature of your glob, in which case this would be 0 work and go back to an `O(1)` loop. Didn't see any, did I miss anything?

I'd not, I could improve it by doing files and dirs separately. This fn can those two as separate args
* For config files, just take the parent
* For config dirs, walk through the ancestry line. That would make it `O(semgrep dir depth)`. I'm assuming that semgrep directories are not deeply nested, and that there are fewer path parts after the config dir than before. That should typically be a non-factor and amortize out to be constant. In my case, we just have a single dir level to keep vendored rules (precisely to avoid unversioned urls).

Reply:
I have an idea to avoid it all - working on it.

Reply:
All right, did a bit of restructure based on the above.

This loop will be _much_ simpler once we deprecate the default option and stop supporting a mix of file configs and folder configs.

Currently, the only issue is that in the default case, I have two separate globs for files and dirs, which
would result in two separate glob expansions. Once we deprecate the default case, that goes away. And a lot of
other bits of logic will get easier too.